jobrunner/sync.py

```diff
@@ -0,0 +1,116 @@
+"""
+Script which polls the job-server endpoint for active JobRequests and POSTs
+back any associated Jobs.
+"""
+import time
+
+import requests
+
+from .log_utils import configure_logging, set_log_context
+from . import config
+from .create_or_update_jobs import create_or_update_jobs
+from .database import find_where
+from .models import JobRequest, Job
+from .string_utils import slugify, project_name_from_url
+
+
+session = requests.Session()
+
+
+def main():
+    while True:
+        sync()
+        time.sleep(config.POLL_INTERVAL)
+
+
+def sync():
+    results = api_get(
+        "job-requests", params={"active": "true", "backend": config.BACKEND}
+    )
+    job_requests = [job_request_from_remote_format(i) for i in results]
+    job_request_ids = [i.id for i in job_requests]
+    for job_request in job_requests:
+        with set_log_context(job_request=job_request):
+            create_or_update_jobs(job_request)
+    jobs = find_where(Job, job_request_id__in=job_request_ids)
+    jobs_data = [job_to_remote_format(i) for i in jobs]
+    api_post("jobs", json=jobs_data)
+
+
+def api_get(*args, **kwargs):
+    return api_request("get", *args, **kwargs)
+
+
+def api_post(*args, **kwargs):
+    return api_request("post", *args, **kwargs)
+
+
+def api_request(method, path, *args, **kwargs):
+    url = "{}/{}".format(config.JOB_SERVER_ENDPOINT.rstrip("/"), path.lstrip("/"))
+    # We could do this just once on import, but it makes changing the config in
+    # tests more fiddly
+    session.auth = (config.QUEUE_USER, config.QUEUE_PASS)
+    response = session.request(method, url, *args, **kwargs)
+    response.raise_for_status()
+    return response.json()
+
+
+def job_request_from_remote_format(job_request):
+    """
+    Convert a JobRequest as received from the job-server into our own internal
+    representation
+    """
+    return JobRequest(
```

Comment:
There have been conversations about enforcing the required contract via an Interface Definition Language. This is possibly worth considering as something that could then be tested in both the server and the runner. 

There is similarly an argument (actually, stronger, I think) for using a type annotation for JobRequest. Though if you're going to bother I think it would then make sense to use annotations everywhere. It would give extra reassurance about the code.  But again, it's moot. Just wanted to throw it out there...